### ISSUE TITLE
Way to see open terminals

### ISSUE THREAD

**Issue by rekendahl** [blinksh/blink#9]
When working with multiple terminals there need to be a way to get a list of what terminals are open. 

My solution would perhaps be that perhaps swiping up with two fingers could raise a tab bar at the bottom (similar to how iTerm) looks. The name of each terminal should (at least) be the name of the machine it's been "moshed" into.

I kind of like the idea that swiping left/right switches tabs and swiping up shows tab list and tap creates new terminal. That still leaves swipe down as a future interaction mechanism

**Comment by carloscabanero** [comment]
I like the idea of swiping up to show the terminals. It also convenient because it would show the list at the bottom, and would be easier to control / closer to the keyboard.

My problem with tab bars or any text lists is that it is difficult to identify them, so you end up just going through all. Maybe a bar with the terminal screenshot could work in this case (or something like safari). 

![apps 11291 9007199266463784 bed8277f-b10d-448a-a14d-b599aaccffdd 88d4b36d-9bc8-4f1e-8679-0065a3d226bb](https://cloud.githubusercontent.com/assets/536729/14265938/d7e18114-fac4-11e5-9ddb-a82118e6f4a4.png)

**Comment by rekendahl** [comment]
I think mimicing iOS large swipe-able cards could work. Might work better if touch is main way of changing tabs (doesn't work as well with cmd+tab). In the picture you show the screenshots for each term seem hard to tell apart. In a small snapshot I think current command, machine name and user is probably most important. 

Hope that helps. Will be interesting to see what you pick.

**Comment by binaryben** [comment]
I obviously haven't seen the UI (yet), but as a potential user, I would appreciate a lean, minimal user interface with a focus on providing a full screen prompt where possible. Other apps often seem to have a bar or something in the way of the prompt interface.

**Comment by carloscabanero** [comment]
We already have a full screen interface, with no space for bars, and worked really hard to keep it like that ;)

I have seen this, that could actually work https://github.com/stefanceriu/SCSafariPageController. It is similar to what Safari does for the tabs, and I kinda like to have that "second" mode, where we could probably show other things like scp progress, notifications...

**Comment by binaryben** [comment]
That is good to hear! I can't wait to try it. Are there any screen shots or forums available to discuss design ideas?

**Comment by carloscabanero** [comment]
Not sure I understand what you mean by keeping the backend stuff on the back? But I think we have a solution.

I tested the cards solution and didn't really like it because it requires you to change your mode everytime you want to switch windows. I didn't even use it much in Safari for navigation anyway. Finally I discovered the problem was location, you didn't really knew where you were after a while, so...

![image](https://cloud.githubusercontent.com/assets/536729/15745545/58bd0cd2-28a0-11e6-8454-f01b227b68e9.png)
![image](https://cloud.githubusercontent.com/assets/536729/15745546/58c752f0-28a0-11e6-8e5f-0d58fa085ccd.png)

**Comment by binaryben** [comment]
Okay, so I've given some thought to how I would like to interact with a mobile shell based on this thread.

## Guestures:

* 1 finger swipe left/right = swipe between terminal sessions with the nice simple HUD UI in your last comment @carloscabanero; this is suitable for "flicking" quickly between sessions
* single finger, single tap = reveal toolbars; see section below
* single finger double tap = open a new session card; when you do this in safari, it zooms in on a web page, so I would expect a new session to have a very simple and clean zoom in transition through the existing session or something similar.
* 2 finger double tap = close the current session; simple enough to perform quickly, complex enough to not accidentally close a tab
* 2 finger pinch would reveal a session switcher similar to the app switcher in iOS than the tab switcher in Safari. I think this is important - assuming it can be updated live - as it means a user could go have a quick status check over multiple sessions (bigger view in the app switcher compared to the linked safari-like solution) without having to completely focus on each session individually.

Again, similar to the iOS app switcher, swipe up on the preview to close the session (don't include a X for closing if possible). Call it the "Session Manager." Ideally, this would have some basic information under each session "frame" as well, again keeping it as minimal as possible, but details like the IP or domain address, maybe a Blink specific nickname and the machine Hostname. It might even be practical to display any live stats if there is anything important a user might want to be able to check quickly (pinch in, check, pinch out)?

## Toolbars:

There is an app called Frameless on the App Store. All it does is provide a webview/browser with the ability to hide the Chrome that comes in Safari. You swipe left and right to go back and forward through history, and up and down to reveal the chrome. Download it and have a look if I make no sense.. :)

Essentially, the overall navigation I am suggesting is similar to that, except left right is a quick selector and a single tap reveals the toolbars as there is less to "tap/click" in a terminal (plus.. I assume single finger swipe up down is scrolling?). I like the first mockup (or is that a live screenshot?) you posted with toolbars, but I think the toolbars are the wrong way round.

The bottom bar should be the translucent bar with commonly used shortcuts (cut/paste + other tools that you might create in the future) and the screenshots you have on the top bar. I would expect the workflow to be: single tab reveals the tool bar, you can then swipe on the main area and switch between sessions like normal except the screenshots at the bottom indicate which is in focus, or you can choose to swipe left and right on the actual screenshots and tap on the session you want in focus.

The top bar should be opaque, and please a status bar! Again, good workflow: tap to check the time and battery status, tap to get back to work. From there you might want to include a settings icon, a [+] icon for creating new sessions. Maybe even a [  ]+ (the box/page with a plus on the outside for creating a new set of credentials/logons).

## Summary:

* Single finger swipes = navigation (left/right as quick selector, up/down as scrolling)
* single finger tap = reveal/hide toolbars
* Double tap = open/close session (one finger for open, two for close)
* Pinch = open/close session manager
* Core navigation and actions should have icon shortcuts available in a hidden toolbar with a preview of open sessions on the bottom toolbar bar.

> I tested the cards solution [...]

Implementing three different ways to navigate (not counting keyboard shortcuts?) would possibly be a lot of replication of basic functionality, but each different way to navigate would likely be used for slightly different purposes in my experience, and makes an app really user friendly. I also think that having larger cards (like the iOS app switcher) allows that method of input to have a greater reason for existence.

I am really sorry if this is already implemented/considered or irrelevant btw. And it is of course just my humble 2¢. Implement all of it, none of it. I am happy either way. Also sorry for the long winded comment - but if you need me to, I can mock up the idea as a web app to demonstrate what I had in mind.

**Comment by BinaryBen** [comment]
I'm adding a request in terms of how the switching gestures are handled, which seems appropriate to this issue.

I like the way one finger left/right switches between open clients and up/down scrolls. When you use two fingers down to close though, it moves the modal HUD like this:

![image](https://cloud.githubusercontent.com/assets/12827556/23008776/f0d5f704-f44c-11e6-9308-0b26a0eae6d8.png)
![image](https://cloud.githubusercontent.com/assets/12827556/23008777/f155bd4a-f44c-11e6-815a-4a8fa664e2b9.png)
![image](https://cloud.githubusercontent.com/assets/12827556/23008778/f1594640-f44c-11e6-805f-c7f5651d330a.png)

It looks very unnatural to me and I'd prefer if the modal HUD (let me know if you have a more official name for this, the grey box with blue dots in it) stayed in a consistent position, and if the terminal itself wasn't resized (and instead was just faded out). Ideally, the client terminal should remain full screen all the time - unless resizing for split screen in future versions of course!

Similarly, the movement of the modal HUD when the keyboard bar appears/disappears feels glitchy when you have a BT keyboard attached. It might be more appropriate to only move it when the entire onscreen keyboard is toggled on and off rather than just the function keys?

Forgive me if it seems picky, UX is my jam! It's the small things that make a (relatively) expensive app a must have for developers though.

**Comment by carloscabanero** [comment]
Man, you hit the spot with this! It was a pain even during SplitView. Here are the fixes I made:
- I have added some tolerance before changing the position of the HUD (the grey modal).
- The HUD will only appear when Blink is firstResponder, to avoid appearing again and again during SplitView.
- When dragging down with two fingers, the terminal won't change size anymore so the experience is smoother and we do not show any HUD.
- Moving the modal now happens animated, so the experience is a whole lot better when rotating the device or when connecting to a keyboard.

I love this UX suggestions, so keep them coming!
